src/data/nat/fib/zeckendorf.lean

```diff
@@ -0,0 +1,229 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakub Kądziołka
+-/
+import data.nat.fib.basic
+import algebra.big_operators.order
+import tactic.omega.main
+import tactic.linarith
+
+/-!
+# Zeckendorf's Theorem
+
+`nat.zeckendorf`: Each natural number can be written in exactly one way as a sum of distinct,
+non-consecutive Fibonacci numbers.
+-/
+
+open_locale big_operators
+
+namespace nat
+
+/-- The set `S` contains the indices which have been chosen.
+    For uniqueness, we need to exclude `fib 0` and `fib 1`,
+    since `fib 0 = 0` and `fib 1 = fib 2 = 1`. -/
+def is_zeckendorf_rep (S : finset ℕ) : Prop :=
+  ∀ (n ∈ S), n ≥ 2 ∧ n + 1 ∉ S
+
+/-- Every natural number lies between two Fibonacci numbers. -/
+lemma fib_between (n : ℕ): ∃ k : ℕ, fib k ≤ n ∧ n < fib k.succ :=
+begin
+  induction n using nat.two_step_induction with n IHn IHnsucc,
+  { use 0, simp },
+  { use 2, simp },
+  { obtain ⟨k, Hfibk_around_nsucc⟩ := IHnsucc,
+    by_cases Hfib_ksucc : n.succ.succ = fib k.succ,
+    { use k.succ,
+      have k_not_0 : k ≠ 0,
+      { assume H : k = 0,
+        have bad := calc n.succ.succ = fib k.succ : Hfib_ksucc
+                                 ... = fib 1      : by simp [H]
+                                 ... = 1          : rfl,
+        cases bad },
+      split,
+      show fib k.succ ≤ n.succ.succ, by simp [Hfib_ksucc],
+      show n.succ.succ < fib k.succ.succ,
+      { rw Hfib_ksucc, apply fib_lt_fib_succ, omega } },
+    { use k, split; omega } }
+end
+
+lemma subset_is_zeckendorf {S T : finset ℕ}
+  (HzT : is_zeckendorf_rep T)
+  (Hsubset : S ⊆ T): is_zeckendorf_rep S :=
+begin
+  assume (n : ℕ) (HnS : n ∈ S),
+  obtain ⟨Hgt2, HnconsecT⟩ : n ≥ 2 ∧ n + 1 ∉ T := HzT n (Hsubset HnS),
+  have HnconsecS : n + 1 ∉ S,
+  { assume (HconsecS : n + 1 ∈ S),
+    have HconsecT : n + 1 ∈ T := Hsubset HconsecS,
+    contradiction },
+  tauto
+end
+
+lemma insert_is_zeckendorf
+  (S : finset ℕ) (Hzeckendorf : is_zeckendorf_rep S)
+  (n : ℕ) (Hle_n : ∀ i ∈ S, i ≤ n):
+  is_zeckendorf_rep (insert (n + 2) S) :=
+begin
+  assume (i : ℕ) (Hi : i ∈ insert (n + 2) S),
+  rw finset.mem_insert at Hi, cases Hi,
+  /- Hi : i = n + 2 -/
+  { rewrite Hi at *,
+    have no_n_plus_3 : n + 3 ∉ S,
+    { assume n_plus_3 : n + 3 ∈ S,
+      have bad : n + 3 ≤ n := Hle_n (n + 3) n_plus_3,
+      linarith },
+    finish },
+  /- Hi : i ∈ S -/
+  { obtain ⟨ge2, nconsec⟩ : i ≥ 2 ∧ i + 1 ∉ S := Hzeckendorf i Hi,
+    have not_new : i + 1 ≠ n + 2,
+    { have le_n : i ≤ n := Hle_n i Hi, linarith },
+    finish }
+end
+
+lemma zeckendorf_bound_base_case
+  (S : finset ℕ) (Hzeckendorf : is_zeckendorf_rep S)
+  (n : ℕ) (Hle_n : ∀ i ∈ S, i ≤ n)
+  (Hlt2 : n < 2):
+  S.sum fib < fib n.succ :=
+begin
+  suffices Hempty : S = ∅, by simp [Hempty, fib_pos],
+  apply finset.eq_empty_of_forall_not_mem,
+  assume (i : ℕ) (Hi : i ∈ S),
+  have ge_2: i ≥ 2 := (Hzeckendorf i Hi).elim_left,
+  have Hbad: i ≤ n := Hle_n i Hi,
+  linarith
+end
```

Comment:
Is there a way of deduplicating this code without making it a separate lemma? If I just do `have ...` before `induction`, it will become an assumption of the induction hypothesis...

Reply:
If you think this lemma is too specific to be useful anywhere else, you could mark it as `private`.

Reply:
Though keep in mind that `private` is used *very* sparingly in mathlib: there are at most 459 uses total.

It's fine having supporting lemmas, and I'd suggest looking for even more lemmas to split apart the longer proofs in this PR if possible.